game-core/src/main/java/games/strategy/triplea/delegate/power/calculator/AaDefenseCombatValue.java

```diff
@@ -0,0 +1,95 @@
+package games.strategy.triplea.delegate.power.calculator;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.attachments.UnitSupportAttachment;
+import java.util.function.Predicate;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Value;
+
+/**
+ * Calculates defense strength and roll for AA/Targeted dice
+ *
+ * <p>This takes into account friendly support, and enemy support
+ */
+@Builder
+@Value
+@Getter(AccessLevel.NONE)
+class AaDefenseCombatValue implements CombatValue {
+
+  @NonNull GameData data;
+  @NonNull AvailableSupports friendlySupportTracker;
+  @NonNull AvailableSupports enemySupportTracker;
+
+  @Override
+  public StrengthOrRollCalculator getRoll() {
+    return new AaDefenseRoll(friendlySupportTracker, enemySupportTracker);
+  }
+
+  @Override
+  public StrengthOrRollCalculator getStrength() {
+    return new AaDefenseStrength(data, friendlySupportTracker, enemySupportTracker);
+  }
+
+  @Override
+  public boolean isDefending() {
+    return true;
+  }
+
+  @Override
+  public GameData getGameData() {
+    return data;
+  }
+
+  static class AaDefenseRoll extends StrengthOrRollCalculator {
+
+    AaDefenseRoll(final AvailableSupports friendlySupport, final AvailableSupports enemySupport) {
+      super(friendlySupport, enemySupport);
+    }
+
+    @Override
+    public int getValue(final Unit unit) {
+      final RollValue rollValue =
+          RollValue.of(unit.getUnitAttachment().getMaxAaAttacks())
+              .add(addSupport(unit, friendlySupportTracker))
+              .add(addSupport(unit, enemySupportTracker));
+      return rollValue.getValue();
+    }
+
+    @Override
+    protected Predicate<UnitSupportAttachment> getRuleFilter() {
```

Comment:
I don't know if a 'rule filter' is cohesive to an API for defense combat value. This seems like it probably should belong to another interface, or a different class. Alternatively, the filter perhaps should accept a collection and then return a filtered result so that the caller is not aware that a predicate is being used. But again, offhand, filtering units seems to be a slightly different operation type compared to combat value calculation. Thoughts on any options around this?

Reply:
I think I can get rid of the rule filter completely.  The initial place where unit support attachments is grabbed uses a filter similar to this.  I plan to look into it more once I've done more restructuring of this code.  So, I think this should stay for now and then later rework how the attachments are grabbed initially to not need this filter here.